weave/trace_server/llm_completion.py

```diff
@@ -17,6 +18,137 @@
+                ) from e
+
+        formatted_messages.append(formatted_message)
+
+    return formatted_messages
+
+
+def resolve_prompt_messages(
+    prompt: str,
+    project_id: str,
+    obj_read_func: Callable[[tsi.ObjReadReq], tsi.ObjReadRes],
+    template_vars: Optional[dict[str, Any]] = None,
+) -> list[dict[str, Any]]:
+    """Resolve a prompt reference to its messages.
+
+    Args:
+        prompt: The prompt reference (e.g., "prompt-name:v2")
+        project_id: The project ID
+        obj_read_func: Function to read objects from the database
+        template_vars: Optional template variables to substitute in messages
+
+    Returns:
+        List of messages from the prompt (with template variables substituted if provided)
+    """
+    # Parse the prompt to extract object_id and digest
+    # Format can be:
+    # 1. Internal URI: "weave-trace-internal:///project_id/object/name:digest"
+    # 2. Full weave URI: "weave:///entity/project/object/name:digest"
+    # 3. Simple format: "object_id:digest" or "object_id:vN"
+
+    if prompt.startswith("weave-trace-internal:///") or prompt.startswith("weave:///"):
+        # Parse using internal URI parser
+        from weave.trace_server import refs_internal as ri
+
+        try:
+            parsed_ref = ri.parse_internal_uri(prompt)
+            if isinstance(parsed_ref, ri.InternalObjectRef):
+                object_id = parsed_ref.name
+                version = parsed_ref.version
+                # Override project_id from the URI if provided
+                if parsed_ref.project_id:
+                    project_id = parsed_ref.project_id
+            else:
+                raise InvalidRequest(
+                    f"Prompt reference must be an object reference, got: {type(parsed_ref)}"
+                )
+        except Exception as e:
+            raise InvalidRequest(
+                f"Failed to parse prompt reference {prompt}: {e!s}"
+            ) from e
+    elif ":" in prompt:
+        object_id, version = prompt.rsplit(":", 1)
+    else:
+        object_id = prompt
+        version = "latest"
+
+    # Fetch the prompt object
+    try:
+        prompt_obj_req = tsi.ObjReadReq(
+            project_id=project_id,
+            object_id=object_id,
+            digest=version,
+            metadata_only=False,
+        )
+        prompt_obj_res = obj_read_func(prompt_obj_req)
+
+        if prompt_obj_res.obj is None:
+            raise NotFoundError(f"Could not find prompt with ref {prompt}")
+
+        # Validate that this is a Prompt or MessagesPrompt object
+        if prompt_obj_res.obj.base_object_class not in ("Prompt", "MessagesPrompt"):
+            raise InvalidRequest(
+                f"Prompt {prompt} is not a Prompt or MessagesPrompt (found {prompt_obj_res.obj.base_object_class})"
+            )
+
+        # Extract messages from the prompt object
+        prompt_val = prompt_obj_res.obj.val
```

Comment:
starting from here down seems like the logic should be shared with the prompt object